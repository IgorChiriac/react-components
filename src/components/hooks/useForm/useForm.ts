import { useState, useCallback, useMemo } from 'react';

type TValue = any;

type TValues = Record<string, TValue>;

interface TUseFormProps {
  initialValues: TValues;
  onSubmit: (values: TValues) => void;
  validate?: (values: TValues) => Record<keyof TValues, string | undefined>;
  submitting?: boolean;
}

const useForm = ({ initialValues, validate, onSubmit, submitting = false }: TUseFormProps) => {
  const [values, updateValues] = useState(initialValues);
  const [errors, updateErrors] = useState(validate ? validate(initialValues) : {});
  const [touched, updateTouched] = useState<Record<string, boolean>>({});

  const runValidation = useCallback(
    formValues => {
      if (validate) {
        updateErrors(validate(formValues));
      }
    },
    [validate],
  );

  const invalid = useMemo(() => !!Object.keys(errors).length, [errors]);

  const getFieldProps = useCallback(
    name => ({
      disabled: submitting,
      error: errors[name],
      touched: !!touched[name],
      value: values[name],
      onChange(e) {
        const newValues = { ...values, [name]: e.target.value };
        updateValues(newValues);
        runValidation(newValues);
      },
      onBlur() {
        updateTouched({ ...touched, [name]: true });
      },
    }),
    [errors, runValidation, submitting, touched, values],
  );

  const handleSubmit = useCallback(
    e => {
      e.preventDefault();
      if (!invalid && onSubmit) {
        onSubmit(values);
      }
    },
    [invalid, onSubmit, values],
  );

  return {
    getFieldProps,
    invalid,
    submitting,
    handleSubmit,
  };
};

export default useForm;
