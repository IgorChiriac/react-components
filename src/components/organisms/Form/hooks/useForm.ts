import { useState, useCallback, useMemo } from 'react';

type TFormValue = any;
export type TFormValues = Record<string, TFormValue>;
export type TFormErrors = Record<string, string | undefined>;
export type TFormTouched = Record<string, boolean>;

export interface IUseFormProps {
  initialValues: TFormValues;
  onSubmit: (values: TFormValues) => void;
  onChange?: (values: TFormValues) => void;
  validate?: (values: TFormValues) => TFormErrors;
}

export interface IFieldProps {
  error: string | undefined;
  touched: boolean;
  value: TFormValue;
  onChange: (eventData: unknown) => void;
  onBlur: (eventData: unknown) => void;
}

export interface IUseFormValues {
  getFieldProps: (name: string) => IFieldProps;
  handleSubmit: (values: TFormValues) => void;
  invalid: boolean;
  errors: TFormErrors;
  touched: TFormTouched;
}

export const useForm = ({ initialValues, validate, onSubmit, onChange }: IUseFormProps): IUseFormValues => {
  const [values, updateValues] = useState(initialValues);
  const [errors, updateErrors] = useState<TFormErrors>(validate ? validate(initialValues) : {});
  const [touched, updateTouched] = useState<TFormTouched>({});

  const runValidation = async (formValues: typeof initialValues) => {
    if (validate) {
      const errors = validate(formValues);
      updateErrors(errors);
    }
  };

  const invalid = useMemo(() => !!Object.keys(errors).length, [errors]);

  const getFieldProps = useCallback(
    (name: string) => ({
      error: errors[name],
      touched: !!touched[name],
      value: values[name],
      onChange(value: unknown) {
        const newValues = { ...values, [name]: value };
        onChange && onChange(newValues);
        updateValues(newValues);
        runValidation(newValues);
      },
      onBlur() {
        updateTouched({ ...touched, [name]: true });
      },
    }),
    [errors, runValidation, touched, values],
  );

  const handleSubmit = useCallback(
    e => {
      e.preventDefault();
      updateTouched(
        Object.keys(initialValues).reduce(
          (acc, key) => ({
            ...acc,
            [key]: true,
          }),
          {},
        ),
      );
      if (!invalid) {
        onSubmit(values);
      }
    },
    [invalid, onSubmit, values],
  );

  return {
    getFieldProps,
    invalid,
    handleSubmit,
    errors,
    touched,
  };
};
