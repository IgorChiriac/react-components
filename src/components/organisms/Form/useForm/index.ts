export * from './useForm';
export { useFieldContext } from './useFieldContext';
export { useFormContext } from './useFormContext';
export { FormContext } from './FormContext';
