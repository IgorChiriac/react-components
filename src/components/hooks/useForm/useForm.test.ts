import { renderHook, act } from '@testing-library/react-hooks';
import useForm, { TValues, TErrors } from './useForm';

const initialValues = {
  name: 'name',
  lastName: 'lastName',
};
const errorMessage = 'required';
const onSubmit = jest.fn();
const validate = (values: TValues) => {
  const errors: TErrors = {};
  if (values.name === '') {
    errors.name = errorMessage;
  }
  if (values.lastName === '') {
    errors.lastName = errorMessage;
  }
  return errors;
};

describe('useForm', () => {
  afterEach(() => {
    onSubmit.mockClear();
  });

  it('should return the correct object', async () => {
    const { result, waitForNextUpdate } = renderHook(() =>
      useForm({
        initialValues,
        onSubmit,
        validate,
      }),
    );
    await waitForNextUpdate();

    ['getFieldProps', 'invalid', 'handleSubmit'].forEach(property => {
      expect(result.current).toHaveProperty(property);
    });
  });

  describe('invalid form flag', () => {
    it('should be false if no validate function is provided', async () => {
      const { result } = renderHook(() =>
        useForm({
          initialValues,
          onSubmit,
        }),
      );
      expect(result.current.invalid).toEqual(false);
    });

    it('should be false if the fields are valid', async () => {
      const { result, waitForNextUpdate } = renderHook(() =>
        useForm({
          initialValues,
          onSubmit,
          validate,
        }),
      );
      await waitForNextUpdate();
      expect(result.current.invalid).toEqual(false);
    });

    it('should be true if at least one of fields is invalid', async () => {
      const { result, waitForNextUpdate } = renderHook(() =>
        useForm({
          initialValues: {
            ...initialValues,
            name: '',
          },
          onSubmit,
          validate,
        }),
      );
      await waitForNextUpdate();
      expect(result.current.invalid).toEqual(true);
    });
  });

  describe('handleSubmit', () => {
    it('should mark all fields as touched', async () => {
      const preventDefault = jest.fn();
      const { result, waitForNextUpdate } = renderHook(() =>
        useForm({
          initialValues,
          onSubmit,
          validate,
        }),
      );
      await waitForNextUpdate();
      expect(result.current.getFieldProps('name').touched).toEqual(false);
      expect(result.current.getFieldProps('lastName').touched).toEqual(false);
      act(() => {
        result.current.handleSubmit({ preventDefault });
      });
      expect(result.current.getFieldProps('name').touched).toEqual(true);
      expect(result.current.getFieldProps('lastName').touched).toEqual(true);
    });

    it('should fire provided onSubmit callback with form values', async () => {
      const preventDefault = jest.fn();
      const { result, waitForNextUpdate } = renderHook(() =>
        useForm({
          initialValues,
          onSubmit,
          validate,
        }),
      );
      await waitForNextUpdate();
      act(() => {
        result.current.handleSubmit({ preventDefault });
      });
      expect(preventDefault).toHaveBeenCalled();
      expect(onSubmit).toHaveBeenCalledWith(initialValues);
    });

    it('should not fire provided onSubmit callback if the form is invalid', async () => {
      const preventDefault = jest.fn();
      const { result, waitForNextUpdate } = renderHook(() =>
        useForm({
          initialValues: {
            ...initialValues,
            name: '',
          },
          onSubmit,
          validate,
        }),
      );
      await waitForNextUpdate();
      act(() => {
        result.current.handleSubmit({ preventDefault });
      });
      waitForNextUpdate();
      expect(onSubmit).not.toHaveBeenCalled();
    });
  });

  describe('onChange', () => {
    it('should be called with the form values when any of them change', async () => {
      const onChange = jest.fn();
      const { result, waitForNextUpdate } = renderHook(() =>
        useForm({
          initialValues,
          onSubmit,
          onChange,
          validate,
        }),
      );
      await waitForNextUpdate();

      const name = 'Jan';
      act(() => {
        result.current.getFieldProps('name').onChange(name);
      });
      await waitForNextUpdate();
      expect(onChange).toHaveBeenCalledWith({
        ...initialValues,
        name,
      });

      const lastName = 'Dzban';
      act(() => {
        result.current.getFieldProps('lastName').onChange(lastName);
      });
      await waitForNextUpdate();
      expect(onChange).toHaveBeenCalledWith({
        name,
        lastName,
      });
    });
  });

  describe('getFieldProps', () => {
    it('should return the correct object', async () => {
      const { result, waitForNextUpdate } = renderHook(() =>
        useForm({
          initialValues,
          onSubmit,
          validate,
        }),
      );
      await waitForNextUpdate();

      ['error', 'touched', 'value', 'onChange', 'onBlur'].forEach(property => {
        expect(result.current.getFieldProps('name')).toHaveProperty(property);
        expect(result.current.getFieldProps('lastName')).toHaveProperty(property);
      });
    });
  });

  describe('onChange', () => {
    it('should properly handle value change', async () => {
      const { result, waitForNextUpdate } = renderHook(() =>
        useForm({
          initialValues,
          onSubmit,
          validate,
        }),
      );
      await waitForNextUpdate();
      expect(result.current.getFieldProps('name')).toMatchObject({
        value: initialValues.name,
        error: undefined,
      });

      act(() => {
        result.current.getFieldProps('name').onChange('test');
      });
      await waitForNextUpdate();
      expect(result.current.getFieldProps('name')).toMatchObject({
        value: 'test',
        error: undefined,
      });

      act(() => {
        result.current.getFieldProps('name').onChange('');
      });
      await waitForNextUpdate();
      expect(result.current.getFieldProps('name')).toMatchObject({
        value: '',
        error: errorMessage,
      });
    });
  });

  describe('onBlur', () => {
    it('should properly handle blur', async () => {
      const { result, waitForNextUpdate } = renderHook(() =>
        useForm({
          initialValues,
          onSubmit,
          validate,
        }),
      );
      await waitForNextUpdate();

      expect(result.current.getFieldProps('lastName').touched).toEqual(false);

      act(() => {
        result.current.getFieldProps('lastName').onBlur();
      });
      expect(result.current.getFieldProps('lastName').touched).toEqual(true);
    });
  });
});
