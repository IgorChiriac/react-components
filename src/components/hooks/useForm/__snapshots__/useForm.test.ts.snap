// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useForm getFieldProps should return the correct object 1`] = `
Object {
  "disabled": false,
  "error": undefined,
  "onBlur": [Function],
  "onChange": [Function],
  "touched": false,
  "value": "name",
}
`;

exports[`useForm getFieldProps should return the correct object 2`] = `
Object {
  "disabled": false,
  "error": undefined,
  "onBlur": [Function],
  "onChange": [Function],
  "touched": false,
  "value": "lastName",
}
`;

exports[`useForm should return the correct object 1`] = `
Object {
  "getFieldProps": [Function],
  "handleSubmit": [Function],
  "invalid": false,
  "submitting": false,
}
`;
